plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/async/AsyncPartsHandler.java

```diff
@@ -142,7 +144,9 @@ private static void uploadPart(
             () -> s3AsyncClient.uploadPart(
                 uploadPartRequest,
                 AsyncRequestBody.fromInputStream(
-                    inputStreamContainer.getInputStream(),
+                    // Buffered stream is needed to allow mark and reset ops during IO errors so that only buffered
+                    // data can be retried instead of retrying whole file by the application.
+                    new BufferedInputStream(inputStreamContainer.getInputStream(), (int) (ByteSizeUnit.MB.toBytes(1) + 1)),
```

Comment:
@vikasvb90 curious who is closing the `BufferedInputStream`? Does not look like `AsyncRequestBody` will do so

Reply:
Thanks @reta for pointing this!
We are wrapping the input stream with BufferedInputStream. Caller of async S3 api should close the provided input stream which means that we won't have any IO resource left open and all subsequent IO ops on the inner stream should result in stream closed errors. And instance of BufferedInputStream should get gc collected.

In remote store cases, we are closing the streams [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/common/blobstore/transfer/RemoteTransferContainer.java#L238)

Reply:
If reference to `BufferedInputStream` is not present then internal buffer should also be eligible for gc. I can add hook for additional cleanup but doesn't look like a leak. I also don't see any rise in JVM in the benchmark if this was the case.

Reply:
This is resource management discipline: once created, the InputStream should be closed.